### ISSUE TITLE
Direct Connect gateway multi-account support

### ISSUE THREAD

**Issue by ewbankkit** [terraform-providers/terraform-provider-aws#8100]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

Direct Connect gateway multi-account support [announced](https://aws.amazon.com/about-aws/whats-new/2019/03/announcing-multi-account-support-for-direct-connect-gateway/).

**Comment by ewbankkit** [comment]
Requires [AWS SDK v1.19.4](https://github.com/aws/aws-sdk-go/releases/tag/v1.19.4):

> Direct Connect gateway enables you to establish connectivity between your on-premise networks and Amazon Virtual Private Clouds (VPCs) in any commercial AWS Region (except in China) using AWS Direct Connect connections at any AWS Direct Connect location. This release enables multi-account support for Direct Connect gateway, with multi-account support for Direct Connect gateway, you can associate up to ten VPCs from any AWS account with a Direct Connect gateway. The AWS accounts owning VPCs and the Direct Connect gateway must belong to the same AWS payer account ID. This release also enables Direct Connect Gateway owners to allocate allowed prefixes from each associated VPCs.

**Comment by heycasey** [comment]
Looks like the functionality that needs to be added is VGW Association Proposal creation, acceptance/rejection, and deletion. More information here: https://docs.aws.amazon.com/directconnect/latest/UserGuide/multi-account-associate-vgw.html

**Comment by ewbankkit** [comment]
The [`aws_dx_gateway_association`](https://www.terraform.io/docs/providers/aws/r/dx_gateway_association.html) resource enhancement for allowed prefixes has its own issue: https://github.com/terraform-providers/terraform-provider-aws/issues/8199.

**Comment by bflad** [comment]
Starting implementation of the following new resources to handle the cross-account pieces. 👍 

* `aws_dx_gateway_association_proposal`
* `aws_dx_gateway_association_proposal_accepter`

**Comment by bflad** [comment]
First half of support submitted: #8320

The second half will likely be based off #8199 (either accepting new `proposal_id` and `vpn_gateway_owner_account_id` arguments in the existing `aws_dx_gateway_association` resource or  reusing some of the updated handling in a new `aws_dx_gateway_association_proposal_accepter` resource).

**Comment by bflad** [comment]
The first half of this support has been merged with the below:

* New Resource: `aws_dx_gateway_association_proposal`
* data-source/aws_dx_gateway: Add `owner_account_id` attribute
* resource/aws_dx_gateway: Add `owner_account_id` attribute

@ewbankkit has offered development support for the second half of this (accepting proposals) later this week (https://github.com/terraform-providers/terraform-provider-aws/pull/8320#issuecomment-486333597). Thanks, @ewbankkit! 😄 Please do reach out if you have any questions or do not have time to complete this.

**Comment by ewbankkit** [comment]
Proposed documentation for the new `aws_dx_gateway_association_proposal_accepter` resource:

# Resource: aws_dx_gateway_association_proposal_accepter

Provides a resource to manage the accepter's side of a Direct Connect Gateway Association Proposal.
This resource accepts a proposal request (created by another AWS account) to attach a virtual private gateway
(owned by the proposal creator's account) to a Direct Connect gateway (owned by the proposal accepter's account).

## Example Usage

```hcl
provider "aws" {
  # Creator's credentials.
}

provider "aws" {
  alias = "accepter"

  # Accepter's credentials.
}

# Creator's side of the proposal.
data "aws_caller_identity" "creator" {}

resource "aws_vpc" "example" {
  cidr_block = "10.255.255.0/28"
}

resource "aws_vpn_gateway" "example" {
  vpc_id = "${aws_vpc.example.id}"
}

resource "aws_dx_gateway_association_proposal" "example" {
  dx_gateway_id               = "${aws_dx_gateway.example.id}"
  dx_gateway_owner_account_id = "${aws_dx_gateway.example.owner_account_id}"
  vpn_gateway_id              = "${aws_vpn_gateway.example.id}"
}

# Accepter's side of the proposal.
resource "aws_dx_gateway" "example" {
  provider = "aws.accepter"

  name            = "example"
  amazon_side_asn = "64512"
}

resource "aws_dx_gateway_association_proposal_accepter" "example" {
  provider = "aws.accepter"

  proposal_id                  = "${aws_dx_gateway_association_proposal.example.id}"
  dx_gateway_id                = "${aws_dx_gateway.example.id}"
  vpn_gateway_owner_account_id = "${data.aws_caller_identity.creator.account_id}"
}
```

## Argument Reference

The following arguments are supported:

* `dx_gateway_id` - (Required) Direct Connect Gateway identifier.
* `proposal_id` - (Required) Direct Connect Gateway Association Proposal identifier.
* `vpn_gateway_owner_account_id` - (Required) AWS Account identifier of the Virtual Gateway associated with the Direct Connect Gateway.
* `override_allowed_prefixes` - (Optional) Prefixes (CIDRs) that override the VPC prefixes (CIDRs) advertised to the Direct Connect gateway.

### Removing `aws_dx_gateway_association_proposal_accepter` from your configuration

The owner of the virtual private gateway can delete the Direct Connect gateway association proposal if it is still pending acceptance.
After an association proposal is accepted, you can't delete it, but you can disassociate the virtual private gateway from the Direct Connect gateway.
Removing a `aws_dx_gateway_association_proposal_accepter` resource from your configuration will remove it
from your statefile and management, **but will not delete the Direct Connect gateway association proposal.**

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

* `id` - Direct Connect Gateway Association Proposal identifier

## Import

Direct Connect Gateway Association Proposal Accepters can be imported using the proposal ID, e.g.

```
$ terraform import aws_dx_gateway_association_proposal_accepter.example ac90e981-b718-4364-872d-65478c84fafe
```

**Comment by ewbankkit** [comment]
OK, now I've started to implement the acceptance half of this and once the proposal is accepted a Direct Connect Gateway Association is created.
The association is visible from both accounts (creator and accepter) so I think I'll change tack here and take the approach of adding `vpn_gateway_owner_account_id` and `proposal_id` to the [`aws_dx_gateway_association`](https://www.terraform.io/docs/providers/aws/r/dx_gateway_association.html) resource.

**Comment by noah-mercado** [comment]
So @ewbankkit per your last comment what would the workflow for this look like? An `aws_dx_gateway_association_proposal` resource happens in Account A, Account B which houses the DX gateway receives the proposal and a `aws_dx_gateway_association_proposal_accepter` resource is deployed to accept, then back in Account A a `aws_dx_gateway_association` resource needs to deploy using the proposal_id as an attribute?

**Comment by ewbankkit** [comment]
@noah-mercado I've changed my mind again 😄, third time lucky.

I'm now leaning towards a new resource `aws_dx_cross_account_gateway_association` which does the proposal acceptance and assumes responsibility for the underlying AWS DX gateway association resource created once the proposal is accepted. The existing `aws_dx_gateway_association` would remain untouched and be used to manage single account gateway associations.

The workflow would be (based on [this diagram](https://docs.aws.amazon.com/directconnect/latest/UserGuide/multi-account-associate-vgw.html)):

#### Account Z

Create `aws_dx_gateway`

#### Account A

Create `aws_vpc`
Create `aws_aws_vpn_gateway`
Create `aws_dx_gateway_association_proposal` using the VGW ID and DXGW ID plus Account Z's ID

#### Account Z

Create `aws_dx_cross_account_gateway_association` using the ProposalID and Account A' ID

**Comment by noah-mercado** [comment]
@ewbankkit Awesome :) . I definitely like that implementation a lot better. The async pattern of `propose -> accept -> associate` cross accounts could have gotten funky

**Comment by ewbankkit** [comment]
Proposed documentation:

# Resource: aws_dx_cross_account_gateway_association

Associates a Direct Connect Gateway with a VGW in another AWS Account. For single account associations, see the [`aws_dx_gateway_association` resource](/docs/provider/aws/r/dx_gateway_association.html).


To create a cross-account association, create an [`aws_dx_gateway_association_proposal` resource](/docs/provider/aws/r/dx_gateway_association_proposal.html)
in the AWS account that owns the VGW and then accept the proposal in the AWS account that owns the Direct Connect Gateway
by creating an `aws_dx_cross_account_gateway_association` resource.

## Example Usage

```hcl
provider "aws" {
  # Creator's credentials.
}

provider "aws" {
  alias = "accepter"

  # Accepter's credentials.
}

# Creator's side of the proposal.
data "aws_caller_identity" "creator" {}

resource "aws_vpc" "example" {
  cidr_block = "10.255.255.0/28"
}

resource "aws_vpn_gateway" "example" {
  vpc_id = "${aws_vpc.example.id}"
}

resource "aws_dx_gateway_association_proposal" "example" {
  dx_gateway_id               = "${aws_dx_gateway.example.id}"
  dx_gateway_owner_account_id = "${aws_dx_gateway.example.owner_account_id}"
  vpn_gateway_id              = "${aws_vpn_gateway.example.id}"
}

# Accepter's side of the proposal.
resource "aws_dx_gateway" "example" {
  provider = "aws.accepter"

  name            = "example"
  amazon_side_asn = "64512"
}

resource "aws_dx_cross_account_gateway_association" "example" {
  provider = "aws.accepter"

  proposal_id                  = "${aws_dx_gateway_association_proposal.example.id}"
  dx_gateway_id                = "${aws_dx_gateway.example.id}"
  vpn_gateway_owner_account_id = "${data.aws_caller_identity.creator.account_id}"
}
```

## Argument Reference

The following arguments are supported:

* `dx_gateway_id` - (Required) The ID of the Direct Connect gateway.
* `proposal_id` - (Required) The ID of the Direct Connect gateway association proposal.
* `vpn_gateway_owner_account_id` - (Required) The ID of the AWS account that owns the VGW with which to associate the gateway.
* `allowed_prefixes` - (Optional) VPC prefixes (CIDRs) to advertise to the Direct Connect gateway. Defaults to the CIDR block of the VPC associated with the Virtual Gateway. To enable drift detection, must be configured.

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

* `id` - The ID of the Direct Connect gateway association resource.
* `dx_gateway_association_id` - The ID of the Direct Connect gateway association.
* `vpn_gateway_id` - The ID of the VGW with which the gateway is associated.

**Comment by ewbankkit** [comment]
After a (hopefully) final round of mind-changing after starting the work to support transit gateways in the `aws_dx_gateway_association` resource, this functionality will be implemented in https://github.com/terraform-providers/terraform-provider-aws/pull/8528 via changes to the `aws_dx_gateway_association` resource rather than creating a new `aws_dx_cross_account_gateway_association` resource.

## Example Usage

```hcl
provider "aws" {
  # Creator's credentials.
}

provider "aws" {
  alias = "accepter"

  # Accepter's credentials.
}

# Creator's side of the proposal.
data "aws_caller_identity" "creator" {}

resource "aws_vpc" "example" {
  cidr_block = "10.255.255.0/28"
}

resource "aws_vpn_gateway" "example" {
  vpc_id = "${aws_vpc.example.id}"
}

resource "aws_dx_gateway_association_proposal" "example" {
  dx_gateway_id               = "${aws_dx_gateway.example.id}"
  dx_gateway_owner_account_id = "${aws_dx_gateway.example.owner_account_id}"
  associated_gateway_id       = "${aws_vpn_gateway.example.id}"
}

# Accepter's side of the proposal.
resource "aws_dx_gateway" "example" {
  provider = "aws.accepter"

  name            = "example"
  amazon_side_asn = "64512"
}

resource "aws_dx_gateway_association" "example" {
  provider = "aws.accepter"

  proposal_id                         = "${aws_dx_gateway_association_proposal.example.id}"
  dx_gateway_id                       = "${aws_dx_gateway.example.id}"
  associated_gateway_owner_account_id = "${data.aws_caller_identity.creator.account_id}"
}
```

**Comment by bflad** [comment]
The second half of this support (accepting proposals) has been released in [version 2.12.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2120-may-24-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
